_posts/2023-06-30-trino-fest-2023-apacheiceberg.md

```diff
@@ -0,0 +1,63 @@
+---
+layout: post
+title: "CDC patterns in Apache Iceberg"
+author: "Ryan Blue, Ryan Duan"
+excerpt_separator: <!--more-->
+image: /assets/blog/trino-fest-2023/ApacheIceberg.png
+show_pagenav: false
+---
+
+Have you ever needed to keep your data all in a table and an efficient way to
+interact with them? Iceberg, an open standard for tables with SQL behavior, is
+exactly what you need. But what makes Iceberg unique is its use of change data
+capture (CDC). Co-creator of Apache Iceberg, Ryan Blue, gave a talk at [Trino Fest 2023]({% post_url
+2023-06-20-trino-fest-2023-recap %}) this past week detailing the trade-offs
+between each pattern that can be used for writing CDC streams into Iceberg
+tables.
+
+<!--more-->
+
+{% youtube GM7EvRc7_is %}
+
+<a class="btn btn-pink btn-md" target="_blank" href="/assets/blog/trino-fest-2023/TrinoFest2023Iceberg.pdf">
+  Check out the slides!
+</a>
+
+## Recap
+
+To begin, what is CDC and why should you use it? CDC is the idea that relational
+or transactional tables are modified, you emit an update stream to keep copies
+in sync so that you capture changes to tables as they happen. As Ryan states,
+"[CDC] is very lightweight on the source database ... rather than being super
+careful with what we run on the database, what we want to do is just make a
+copy of it very easily and maintain that copy." Ryan continues giving an
+example of a bank using a transactional table in Iceberg to offer some
+context on what's going on.
+
+With this knowledge, you can see some problems that makes CDC difficult:
+
+* Lower latency means more work
+* Write amplification - the work necessary to balance the trade-off between
+  efficiency at write time and efficiency at read time
+* Batch writes with double update and possible inconsistency
+* Read requirements with the different types of deletes in a table
+
+With these types of problems, the importance of the trade-offs between the
+different patterns rises due to the need for utmost efficiency. The first
+trade-off that Ryan talks about is the storage trade-offs between using direct
+writes and a change log table, which is considered the most important and often
+overlooked decision. The next trade-off is in regards to the MERGE pattern's
+choice of lazy merge (merge-on-read) or eager merge (copy-on-write). In
+addition, the commit frequency trade-off has different benefits depending if you
+prefer it to be faster or slower. The Flink UPSERT pattern is an update type
+trade-off, but is not recommended due to its many drawbacks. The change log
+pattern and MERGE pattern both have benefits you may want, so Ryan suggests
+using a hybrid version of both that may give you what you want from both
+patterns. All of these different options create numerous variations depending on
+what you prefer.
+
```

Comment:
Do we need some sort of "watch the video and slides for more details" @colebow  .. I kinda of think maybe yes.. but also maybe not .. 

How about just a final closure sentence?

Reply:
Maybe something like

He mentions that with Iceberg you have the choice and the different CDC patterns can be supported and you can adjust your usage to your specific needs. Watch the video and review the slides for more details.